### ISSUE TITLE
Lints for functions with `Result` return types which only return `Ok(...)` or `Err(...)`

### ISSUE THREAD

**Issue by anp** [rust-lang/rust-clippy#12119]
### What it does

I think it could be useful to have a lint that warned authors that their function didn't need to have a fallible return type because one or other variant would always be returned. The `Result` type is usually used to indicate fallibility, but sometimes in the course of refactoring code failure modes will disappear and authors may miss that the function's signature could be updated to an infallible one.

### Advantage

 Removing unnecessary results should:

* reduce cognitive overhead for callers
* reduce compile times by removing unneeded branches
* improve readability by removing unnecessary error handling code
* more clearly convey behavior of the called code

### Drawbacks

The main drawback I see is the potential for false positives or excessive `#[allow(...)]`s.

I'm not sure whether such a lint should be raised on trait method implementations for exported traits (since some implementations may still need to be fallible). I think private traits where all implementations are known would probably be useful to warn on.

I'm not sure whether such a lint should be raised on publicly exported functions which would require semver breaking changes to update. IMO it's probably OK to report these for for authors to leave an `#[allow(...)]` but I think it bears discussion.

### Example

```rust
fn hello_world() -> Result<String, String> {
    Ok("Hello, world!".to_string())
}
```

Could be written as:

```rust
fn hello_world() -> String {
    "Hello, world!".to_string()
}
```

**Comment by samueltardieu** [comment]
This lint already exists ([`unnecessary-wraps`](https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_wraps)) but you have to opt-in as it is not enabled by default:

```bash
$ cargo clippy -- -D clippy::unnecessary-wraps
error: this function's return value is unnecessarily wrapped by `Result`
  |
1 | / fn hello_world() -> Result<String, String> {
2 | |     Ok("Hello, world!".to_string())
3 | | }
  | |_^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_wraps
  = note: requested on the command line with `-D clippy::unnecessary-wraps`
help: remove `Result` from the return type...
  |
1 | fn hello_world() -> std::string::String {
  |                     ~~~~~~~~~~~~~~~~~~~
help: ...and then change returning expressions
  |
2 |     "Hello, world!".to_string()
```

**Comment by anp** [comment]
Ah, that's great, thank you! I didn't see it when searching the lints list but I don't remember what terms I looked for so not sure whether there's a good way to make it more visible.

**Comment by heiner** [comment]
It's not 100% the same as [unnecessary-wraps](https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_wraps) as that checks functions that "only return `Ok` or `Some`", but not functions that _only_ return `Err` and never `Ok`.
